llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp

```diff
@@ -1235,63 +1244,146 @@ IRBuilder<>::InsertPoint OpenMPIRBuilder::createParallel(
+    OI.PostOutlineCB = [=](Function &OutlinedFn) {
+      // Add some known attributes.
+      OutlinedFn.addParamAttr(0, Attribute::NoAlias);
+      OutlinedFn.addParamAttr(1, Attribute::NoAlias);
+      OutlinedFn.addFnAttr(Attribute::NoUnwind);
+      OutlinedFn.addFnAttr(Attribute::NoRecurse);
+
+      assert(OutlinedFn.arg_size() >= 2 &&
+             "Expected at least tid and bounded tid as arguments");
+      unsigned NumCapturedVars =
+          OutlinedFn.arg_size() - /* tid & bounded tid */ 2;
+
+      CallInst *CI = cast<CallInst>(OutlinedFn.user_back());
+      CI->getParent()->setName("omp_parallel");
+      Builder.SetInsertPoint(CI);
+
+      // Build call __kmpc_fork_call[_if](Ident, n, microtask, var1, .., varn);
+      Value *ForkCallArgs[] = {
+          Ident, Builder.getInt32(NumCapturedVars),
+          Builder.CreateBitCast(&OutlinedFn, ParallelTaskPtr)};
+
+      SmallVector<Value *, 16> RealArgs;
+      RealArgs.append(std::begin(ForkCallArgs), std::end(ForkCallArgs));
+      if (IfCondition) {
+        Value *Cond = Builder.CreateSExtOrTrunc(
+            IfCondition, Type::getInt32Ty(M.getContext()));
+        RealArgs.push_back(Cond);
+      }
+      RealArgs.append(CI->arg_begin() + /* tid & bound tid */ 2, CI->arg_end());
+
+      // __kmpc_fork_call_if always expects a void ptr as the last argument
+      // If there are no arguments, pass a null pointer.
+      auto PtrTy = Type::getInt8PtrTy(M.getContext());
+      if (IfCondition && NumCapturedVars == 0) {
+        llvm::Value *Void = ConstantPointerNull::get(PtrTy);
+        RealArgs.push_back(Void);
+      }
+      if (IfCondition && RealArgs.back()->getType() != PtrTy)
+        RealArgs.back() = Builder.CreateBitCast(RealArgs.back(), PtrTy);
 
-    Builder.CreateCall(RTLFn, RealArgs);
+      Builder.CreateCall(RTLFn, RealArgs);
 
-    LLVM_DEBUG(dbgs() << "With fork_call placed: "
-                      << *Builder.GetInsertBlock()->getParent() << "\n");
+      LLVM_DEBUG(dbgs() << "With fork_call placed: "
+                        << *Builder.GetInsertBlock()->getParent() << "\n");
 
-    InsertPointTy ExitIP(PRegExitBB, PRegExitBB->end());
+      InsertPointTy ExitIP(PRegExitBB, PRegExitBB->end());
 
-    // Initialize the local TID stack location with the argument value.
-    Builder.SetInsertPoint(PrivTID);
-    Function::arg_iterator OutlinedAI = OutlinedFn.arg_begin();
-    Builder.CreateStore(Builder.CreateLoad(Int32, OutlinedAI), PrivTIDAddr);
+      // Initialize the local TID stack location with the argument value.
+      Builder.SetInsertPoint(PrivTID);
+      Function::arg_iterator OutlinedAI = OutlinedFn.arg_begin();
+      Builder.CreateStore(Builder.CreateLoad(Int32, OutlinedAI), PrivTIDAddr);
 
-    CI->eraseFromParent();
+      CI->eraseFromParent();
 
-    for (Instruction *I : ToBeDeleted)
-      I->eraseFromParent();
-  };
+      for (Instruction *I : ToBeDeleted)
+        I->eraseFromParent();
+    };
```

Comment:
ToBeDeleted (and CI->erase) are the same, maybe more. Let’s sink them instead of duplicating them.

Reply:
Fixed